### PR TITLE
Require type hints in the handlers module

### PR DESCRIPTION
**Pull request by clokep**
Similar to what I did in #10736 and then #10817, this requires all methods to be typed in the `synapse.handlers` module.

In order to do that I had to add a bunch of type hints.

This is reviewable commit-by-commit.

### CODE REVIEW

File: synapse/handlers/cas.py [hunk 1]
```diff
@@ -133,11 +133,9 @@ async def _validate_ticket(
...
         except HttpResponseException as e:
             description = (
-                (
-                    'Authorization server responded with a "{status}" error '
-                    "while exchanging the authorization code."
-                ).format(status=e.code),
-            )
+                'Authorization server responded with a "{status}" error '
+                "while exchanging the authorization code."
+            ).format(status=e.code)
```
Comment (clokep) [inline synapse/handlers/cas.py:138]: This was a real bug, it was passing a tuple instead of a string.

File: synapse/handlers/appservice.py [hunk 2]
```diff
@@ -116,9 +116,9 @@ async def handle_event(event: EventBase) -> None:
 
                         if not self.started_scheduler:
 
-                            async def start_scheduler():
+                            async def start_scheduler() -> None:
                                 try:
-                                    return await self.scheduler.start()
+                                    await self.scheduler.start()
```
Comment (clokep) [inline synapse/handlers/appservice.py:121]: `Scheduler.start` doesn't return anything so no need to pass that back.

File: synapse/handlers/sync.py [hunk 3]
```diff
@@ -1534,12 +1534,12 @@ async def _generate_sync_entry_for_rooms(
...
         newly_left_rooms = room_changes.newly_left_rooms
 
-        async def handle_room_entries(room_entry: "RoomSyncResultBuilder"):
+        async def handle_room_entries(room_entry: "RoomSyncResultBuilder") -> None:
             with start_active_span("generate_room_entry"):
                 set_tag("room_id", room_entry.room_id)
                 log_kv({"events": len(room_entry.events or [])})
                 logger.debug("Generating room entry for %s", room_entry.room_id)
-                res = await self._generate_room_entry(
+                await self._generate_room_entry(
```
Comment (clokep) [inline synapse/handlers/sync.py:1542]: _generate_room_entry modifies the RoomSyncResultBuilder and doesn't return anything.

File: synapse/handlers/federation.py [hunk 4]
```diff
@@ -1262,11 +1262,11 @@ def sort_fun(ev: EventBase) -> Tuple[int, str]:
 
         logger.debug("construct_auth_difference before get_next!")
 
-        def get_next(it, opt=None):
```
Comment (clokep) [inline synapse/handlers/federation.py:1265]: `opt` was never passed, it seemed clearer to just inline it below.

File: synapse/handlers/federation.py [hunk 5]
```diff
@@ -1305,6 +1308,7 @@ def get_next(it, opt=None):
             if current_local.event_id < current_remote.event_id:
                 missing_locals.append(current_local)
                 current_local = get_next(local_iter)
+                continue
```
Comment (clokep) [inline synapse/handlers/federation.py:1311]: I believe this is a real bug, without this the below code can run, but with a new `current_local` (which could be `None`).

File: synapse/handlers/account_data.py [hunk 6]
```diff
@@ -171,7 +171,7 @@ def get_current_key(self, direction: str = "f") -> int:
         return self.store.get_max_account_data_stream_id()
 
     async def get_new_events(
-        self, user: UserID, from_key: int, **kwargs
+        self, user: UserID, from_key: int, **kwargs: Any
```
Comment (reivilibre) [inline synapse/handlers/account_data.py:174]: what's this `kwargs` for?

File: synapse/handlers/federation.py (hunk 5 above)
Comment (reivilibre) [reply synapse/handlers/federation.py:1311]: unchecked since I don't know the code well

File: synapse/handlers/receipts.py [hunk 7]
```diff
@@ -216,7 +216,7 @@ def filter_out_hidden(events: List[JsonDict], user_id: str) -> List[JsonDict]:
         return visible_events
 
     async def get_new_events(
-        self, from_key: int, room_ids: List[str], user: UserID, **kwargs
+        self, from_key: int, room_ids: List[str], user: UserID, **kwargs: Any
```
Comment (reivilibre) [inline synapse/handlers/receipts.py:219]: also the `kwargs` here

File: synapse/handlers/typing.py [hunk 8]
```diff
@@ -485,7 +485,7 @@ async def get_new_events_as(
             return (events, handler._latest_room_serial)
 
     async def get_new_events(
-        self, from_key: int, room_ids: Iterable[str], **kwargs
+        self, from_key: int, room_ids: Iterable[str], **kwargs: Any
```
Comment (reivilibre) [inline synapse/handlers/typing.py:488]: also the kwargs here

**Review by reivilibre — COMMENTED** [review]
Everything looks good except:

- Can't verify the bug — would want someone who knew what they were looking at to confirm (though should we consider a test case to confirm?)
- these random `kwargs` — what are they doing there? Can we get rid of them? It seems like they might hide passed-but-unused arguments in the rest of the code, and since they're not being used anyway (as far as I can tell)…

File: synapse/handlers/account_data.py (hunk 6 above)
Comment (clokep) [reply synapse/handlers/account_data.py:174]: I think that `get_new_events` generally takes some random arguments which most things don't care about. I did verify that there *are* other arguments going into this call, I'll double check if we can tighten this up.

Comment (clokep) [reply synapse/handlers/account_data.py:174]: I think this can be cleaned up, but I'd prefer to do it in a separate PR if that's OK!

Comment (clokep) [reply synapse/handlers/account_data.py:174]: I resolved the other threads about this to keep the conversation in one spot.

Comment (reivilibre) [reply synapse/handlers/account_data.py:174]: Sure, take this to another PR if you prefer.

**Comment by erikjohnston** [comment]
FYI this now has conflicts

File: synapse/handlers/federation.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/handlers/federation.py:1311]: `construct_auth_difference` doesn't seem to be used by anything anymore, so let's just delete this entire function?

**Review by erikjohnston — APPROVED** [review]
Otherwise lgtm

File: synapse/handlers/federation.py (hunk 5 above)
Comment (clokep) [reply synapse/handlers/federation.py:1311]: Good catch!

**Pull request merged**